### PR TITLE
Allow passing `None` as parameters of `Fusion.__call__`

### PR DESCRIPTION
**Pull request by asi1024**
```
>>> import cupy
>>>
>>> @cupy.fuse()
... def f(x, y):
...     if y is None:
...         x[:] = -x
...     else:
...         x += y
...
>>> x = cupy.array([1, 2, 3])
>>> x
array([1, 2, 3])
>>> f(x, 1)
>>> x
array([2, 3, 4])
>>> f(x, None)
>>> x
array([-2, -3, -4])
```

### CODE REVIEW

File: cupy/core/fusion.pyx [hunk 1]
```diff
@@ -849,16 +857,18 @@ class Fusion(object):
...
 
-    def _is_cupy_data(self, a):
-        return isinstance(a, (core.ndarray, numpy.generic))
-
     def __call__(self, *args):
         # Inner function of composition of multiple fused functions.
         if _is_fusing():
             return self.func(*args)
 
         # No cupy ndarray exists in the arguments
```
Comment (hvy) [inline cupy/core/fusion.pyx]: Do you think we should move this comment to the early return below?

File: cupy/core/fusion.pyx [hunk 2]
```diff
@@ -729,15 +730,30 @@ class _FusionHistory(object):
...
+                python_var = _FusionVarArray(cuda_var, a.ndim, False)
+            elif a is None:
+                cuda_var = None
+                python_var = None
+            else:
+                cuda_var = self._fresh_premap_param(numpy.dtype(type(a)))
+                python_var = _FusionVarScalar(cuda_var, -1, False)
+            return cuda_var, python_var
+
+        self.ndim = max([a.ndim for a in args if isinstance(a, core.ndarray)])
```
Comment (hvy) [inline cupy/core/fusion.pyx]: Should we change the name of the input argument in the docs from `in_types` to `args` as well?

File: cupy/core/fusion.pyx [hunk 3]
```diff
@@ -714,30 +715,45 @@ class _FusionHistory(object):
...
-        in_ndims = [d for t, d in in_params_info]
-        self.ndim = max(in_ndims)
-        in_params = [self._fresh_premap_param(t) for t in in_dtypes]
-        in_pvars = [_FusionVarScalar(v, d, False)
-                    if d == -1
-                    else _FusionVarArray(v, d, False)
-                    for v, d in zip(in_params, in_ndims)]
-        return_value = func(*in_pvars)
+
+        def gen_abstracted_args(a):
```
Comment (hvy) [inline cupy/core/fusion.pyx]: Although locally used, how about defining it as a regular method, in favor of performance.

File: cupy/core/fusion.pyx [hunk 4]
```diff
@@ -869,17 +879,39 @@ class Fusion(object):
...
+            elif isinstance(arg, float):
+                params_info.append('d')
+            elif isinstance(arg, int):
+                params_info.append('l')
+            elif isinstance(arg, bool):
+                params_info.append('?')
+            elif isinstance(arg, complex):
+                params_info.append('D')
+            else:
+                raise Exception('Does not reach thie line')
```
Comment (hvy) [inline cupy/core/fusion.pyx]: Should we assert false, if it's an implementation error that is?

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Can you check the Jenkins failure? It seems that a test is failing only with Py2, where only CuPy raises an error.

**Comment by asi1024** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Jenkins failure seems unrelated. LGTM!

**Pull request merged**